### PR TITLE
H-2572: Give research agent ability to ask clarifying questions

### PR DESCRIPTION
**Pull request by CiaranMn**
## 🌟 What is the purpose of this PR?

This PR give the coordinating agent in the 'research' action the ability to ask clarifying questions of the user, which the user can answer by clicking on the prompt in the relevant step in the flow.

The agent is prompted to considering whether questions are required at the outset of the flow – while it is doing so, the step UI shows a message indicating that it is deciding whether to ask questions, so that the user is given a hint to stick around until the initial plan has been produced when the agent has no further initial quesitons, at which point the 'deciding whether to ask questions' message disappears.

The agent can ask clarifying questions later in the course of its research also.

## 🔍 What does this change?

As well as the backend and UI changes necessary to support seeking user input:
- Adds a local 'Flow run started' message to the activity log as soon as the flow is started, to avoid an empty activity log
- Adds a  `CreatedPlan` progress log from the coordinating agent, which is how we know when the research agent has no further immediate questions for the user at the outset of the task

### 🚢 Has this modified a publishable library?

This PR:

- [x] does not modify any publishable blocks or libraries, or modifications do not need publishing

### 📜 Does this require a change to the docs?

The changes in this PR:

- [x] are internal and do not require a docs change

### 🕸️ Does this require a change to the Turbo Graph?

The changes in this PR:

- [x] do not affect the execution graph

## ⚠️ Known issues

- As noted in a `@todo`, human input might take a long time to arrive. Therefore we ideally would refactor the Flow steps into child workflows, with every tool call available to agents an individual activity, so that waiting for human response can be awaited at the workflow level and the workflow put to sleep until the response is received. This allows us to avoid waiting in the activity, which means the workflow is still consuming resources with information in memory until the response is received.

## 🛡 What tests cover this?

- None yet

## ❓ How to test this?

1. Start a research flow with vague instructions

## 📹 Demo

https://github.com/hashintel/hash/assets/37743469/b34afc9d-ed90-4dd9-8157-666ae24a6217

### CODE REVIEW

File: apps/hash-ai-worker-ts/src/activities/flow-activities/shared/get-answers-from-human.ts [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+import { requestExternalInput } from "../../shared/request-external-input";
+
+export const getAnswersFromHuman = async (
+  questions: string[],
+): Promise<string> => {
+  const {
+    data: { answers },
+  } = await requestExternalInput({
+    requestId: generateUuid(),
+    stepId: Context.current().info.activityId,
```
Comment (benwerner01) [inline apps/hash-ai-worker-ts/src/activities/flow-activities/shared/get-answers-from-human.ts:13]: The `stepId` can be obtained from `getFlowContext` now.

File: apps/hash-ai-worker-ts/src/activities/flow-activities/research-entities-action/open-ai-coordinating-agent.ts [hunk 2]
```diff
@@ -181,38 +184,48 @@ const getNextToolCalls = async (params: {
...
 
   const systemPrompt = dedent(`
     ${generateSystemPromptPrefix({ input })}
-
-    Do not make *any* tool calls. You must first provide a plan of how you will use
-      the tools to progress towards completing the task.
-
-    This should be a list of steps in plain English.
+    
+    ${questionsAndAnswers ? `You previously asked the user clarifying questions on the research brief, and received the following answers:\n${questionsAndAnswers}` : ""}
```
Comment (benwerner01) [inline apps/hash-ai-worker-ts/src/activities/flow-activities/research-entities-action/open-ai-coordinating-agent.ts:203]: The question answer pairs are only going to be visible in the system prompt when creating the plan, is this intentional?

I would've expected that the question answer pairs get added to the `state` of the coordinator agent, and then used in the `generateSystemPromptPrefix` method so that they're in context when creating the initial plan, and when the agent is making subsequent tool calls.

File: apps/hash-frontend/src/pages/[shortname]/flows/[flow-uuid].page/flow-definition/swimlane/custom-node.tsx [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+    statusData,
+  });
+
+  const outstandingInputRequest = useMemo(() => {
+    /**
+     * We'll take the first human input request that is not resolved,
+     * and if none are found return the first unresolved browser plugin request.
+     *
+     * There should only ever be a single human input request active,
+     * as they are asked by the top-level co-ordinating agent.
```
Comment (benwerner01) [inline apps/hash-frontend/src/pages/[shortname]/flows/[flow-uuid].page/flow-definition/swimlane/custom-node.tsx:96]:
```suggestion
     * as they are asked by the top-level coordinating agent.
```

**Review by benwerner01 — CHANGES_REQUESTED** [review]
Thank you Ciaran, the changes look good 👍 

Have left some comments mostly related to making the changes in the research activity more consistent with prior work.

I've not been able to test this locally, when running the "Research Entities" the following error is encountered in the HASH API when polling for the flow status:
```
@apps/hash-api:dev: error: graphql {"elapsed":"10.06ms","errors":[{"locations":[{"column":7,"line":16}],"message":"Cannot return null for non-nullable field StepRun.status.","path":["getFlowRuns",0,"steps",0,"status"]}],"operation":"getFlowRuns","requestId":"EWOBHYPLSOuZxr","service":"api","stack":["Error: Cannot return null for non-nullable field StepRun.status.\n    at completeValue (/Users/benwerner/Documents/HASH/hash/node_modules/graphql/execution/execute.js:605:13)\n    at executeField (/Users/benwerner/Documents/HASH/hash/node_modules/graphql/execution/execute.js:500:19)\n    at executeFields (/Users/benwerner/Documents/HASH/hash/node_modules/graphql/execution/execute.js:414:22)\n    at completeObjectValue (/Users/benwerner/Documents/HASH/hash/node_modules/graphql/execution/execute.js:925:10)\n    at completeValue (/Users/benwerner/Documents/HASH/hash/node_modules/graphql/execution/execute.js:646:12)\n    at completeValue (/Users/benwerner/Documents/HASH/hash/node_modules/graphql/execution/execute.js:595:23)\n    at /Users/benwerner/Documents/HASH/hash/node_modules/graphql/execution/execute.js:707:25\n    at Function.from (<anonymous>)\n    at completeListValue (/Users/benwerner/Documents/HASH/hash/node_modules/graphql/execution/execute.js:687:34)\n    at completeValue (/Users/benwerner/Documents/HASH/hash/node_modules/graphql/execution/execute.js:618:12)"],"timestamp":"2024-05-14T08:54:42.920Z","userAgent":"Mozilla/5.0"}
```

Just incase it may have been caused by the polling for some other flow I ran on a different branch, I've reseeded my db but am still encountering the error so there's possibly something that's not being accounted for when determining the step run's status.

File: apps/hash-ai-worker-ts/src/activities/flow-activities/shared/get-answers-from-human.ts (hunk 1 above)
Comment (CiaranMn) [reply apps/hash-ai-worker-ts/src/activities/flow-activities/shared/get-answers-from-human.ts]: Done in https://github.com/hashintel/hash/pull/4461/commits/0cce668ff26764f8536f632d796bf73a594e1b68

File: apps/hash-ai-worker-ts/src/activities/flow-activities/research-entities-action/coordinator-tools.ts [hunk 4]
```diff
@@ -35,6 +36,26 @@ export const coordinatorToolDefinitions: Record<
...
+          items: {
+            type: "string",
+            description:
+              "A question to help clarify or complete the research task",
+          },
+          description: "A list of questions to ask the user",
+        },
+      },
+      required: ["questions"],
+    },
```
[the remark this answers is not in the document]
Comment (CiaranMn) [reply apps/hash-ai-worker-ts/src/activities/flow-activities/research-entities-action/coordinator-tools.ts]: Done in https://github.com/hashintel/hash/pull/4461/commits/6787a4e4f68d262fa60786fbcebb76a241efb097

**Review by benwerner01 — APPROVED** [review]
